CefSharp/ModelBinding/TypeSafeBinder.cs

```diff
@@ -0,0 +1,202 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+using CefSharp.ModelBinding.Converters;
+
+namespace CefSharp.ModelBinding
+{
+
+    /// <summary>
+    /// This class is responsible for marshaling Javascript objects into their corresponding .NET type. 
+    /// </summary>
+    /// <remarks>
+    /// This binder has no backwards compatibility with the <see cref="DefaultBinder"/> due to changes in how data member are marshaled.
+    /// </remarks>
+    public class TypeSafeBinder : IBinder
+    {
+        /// <summary>
+        /// Used to try and convert a generic type to an array via Reflection.
+        /// </summary>
+        private static readonly MethodInfo ToArrayMethodInfo = typeof(Enumerable).GetMethod("ToArray", BindingFlags.Public | BindingFlags.Static);
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="TypeSafeBinder"/> and registers converters with <see cref="TypeDescriptor"/> <br/>
+        /// This binder can reliably marshal data between the Javascript and .NET domains. <br/>
+        /// For example it provides interoperability for Typescript, Javascript, and C# style conventions without making you drop conventions in one of your languages. <br/>
+        /// It will also handle types like <see cref="Guid"/> or <see cref="Enum"/> fields that have flags. <br/>
+        /// Finally it ensure there is type safety by throwing <see cref="TypeBindingException"/> whenever data is malformed so you can catch issues in your code.
+        /// </summary>
+        public TypeSafeBinder()
+        {
+            BinderGuidConverter.Register();
+            BinderVersionConverter.Register();
+        }
+
+        /// <summary>
+        /// Attempts to bind a Javascript object into a corresponding .NET type.
+        /// </summary>
+        /// <param name="javaScriptObject">An instance of the Javascript object.</param>
+        /// <param name="nativeType">The type this method will try to create an instance of using the Javascript object.</param>
+        /// <returns>An instance of the native type.</returns>
+        public object Bind(object javaScriptObject, Type nativeType)
+        {
+            // if the intended destination is an enumeration, we can try and get the corresponding member upfront.
+            // internally this will throw a TypeBindingException if it runs into issues. See the documentation for more information.
+            if (nativeType.IsEnum)
+            {
+                return nativeType.CreateEnumMember(javaScriptObject);
+            }
+
+            // if the source object is null and is not an enum, then there isn't anything left to do.
```

Comment:
We should probably make this a configurable option then. For example in our production use we ran into a bug and didn't realize the Javascript object was null because it was using the default of the Enum. There is some value in strict null checks from a development perspective.

Reply:
Sorry, somehow I totally missed this set of comments.

Configurable is fine with me. I can see people wanting to use either option.